### PR TITLE
Feature/bind by type handling generic

### LINKED ISSUE #109: feature / handle generics (actually GenericAlias) in bind_by_type
For now, the support of generics is not guaranteed.

`typing.Generic` do [caching](https://github.com/python/cpython/blob/3079aa3f5d4f54a05b9b67cb60625e45581f7ade/Lib/typing.py#L1803-L1850).
But cache size is limited.
When new instances of `GenericAlias` are being created - the current hook in `bind_by_type` does not handle this case because it does the `is` check.
https://github.com/adriangb/di/blob/e88c8f9f1e6ddd06d41095025ea209f6f8d15056/di/_container.py#L81-L98

I propose to do a `==` check.
It is completely safe for simple types ([default `__eq__` is actualy `is`](https://docs.python.org/3/reference/datamodel.html?highlight=By%20default,%20object%20implements%20__eq__()%20by%20using%20is#object.__eq__)).
But also this will handle [GenericAlias `__eq__`](https://github.com/python/cpython/blob/3079aa3f5d4f54a05b9b67cb60625e45581f7ade/Lib/typing.py#L1365-L1369).

### PR DESCRIPTION
**Pull request by maxzhenzhera**
Closes #109

### CODE REVIEW

File: .github/workflows/workflow.yaml [hunk 1]
```diff
@@ -42,7 +42,9 @@ jobs:
             os: ubuntu-latest
           - python: "3.10"
             os: ubuntu-latest
-          - python: "3.11.0-beta.1 - 3.11"
+          - python: "3.11"
+            os: ubuntu-latest
+          - python: "3.12"
```
Comment (adriangb) [inline .github/workflows/workflow.yaml:47]: 👍🏻

**Review by adriangb — APPROVED** [review]

**Comment by adriangb** [comment]
Please fix the tests / use `typing_extensions` imports then LGTM

**Comment by maxzhenzhera** [comment]
> Please fix the tests / use typing_extensions imports then LGTM

Omg, forgot about it, few seconds...

**Comment by adriangb** [comment]
> Also:
> 
> * I made a small change in `pyproject.toml` - moved last added deps to categories.
> * Added `python3.12` to ci test job

Yeah I'm good with that and generally don't mind minor stuff like this being sneaked into a PR. Keyword minor.

> And I have a question: I wanted to add a `python3.12` to the badge of supported versions but investigated that this badge is pulled using PyPi. In `pyproject` classifiers list I did not find a list of supported versions.
> 
> Can you tell me, please, how the list of supported python versions is being updated?

I honestly am not sure / don't remember. Let's publish a new version and see if it just updates.

> And about versions: maybe it is time to remove `python3.7` from **ci**?

That said if we're also going to remove 3.7 it seems worth it to make a separate PR to drop 3.7 and add 3.12, so mind moving that from this PR to a new one?

**Comment by maxzhenzhera** [comment]
[fb11e768637f8783da97cc4b25e538844e52f3fb](https://github.com/adriangb/di/pull/110/commits/fb11e768637f8783da97cc4b25e538844e52f3fb) - to test it **cache cleaning** needed and I don't know a simpler way to do it.

Here I marked a new test case to skip for versions that do not allow passing `not types` to `Generic`.

**Comment by maxzhenzhera** [comment]
> That said if we're also going to remove 3.7 it seems worth it to make a separate PR to drop 3.7 and add 3.12, so mind moving that from this PR to a new one?

It is cool.
But I don't know what is the best way to move that commits to another PR?
`git reset` here and make those changes in the new PR?

**Comment by adriangb** [comment]
I would just manually copy pasta the code changes, everything will be squash merged anyway.

File: tests/test_binding.py [hunk 2]
```diff
@@ -47,6 +49,63 @@ def __init__(self, v: int = 1) -> None:
...
+    hook = bind_by_type(
+        Dependent(factory),
+        GetterInterface[int],
+    )
+    container.bind(hook)
+
+    # ===========================================
+    # clean `_tp_cache`
+    lru_cache_size = 128
+    _ = [GetterInterface[i] for i in range(lru_cache_size)]  # type: ignore[valid-type]
```
Comment (adriangb) [inline tests/test_binding.py:88]: You could just generate 128+ types by parametrizing, something like:

```python
unique_types = {
    Tuple[*perm] for n in range(1, 8)
    for comb in itertools.combinations([str, int, float, bool, None, object, Any], r=n)
    for perm in itertools.permutations(comb)
}
```

**Comment by maxzhenzhera** [comment]
Not to do a `force` push I just deleted 3.12 from the workflow in the last commit.

File: tests/test_binding.py (hunk 2 above)
Comment (maxzhenzhera) [reply tests/test_binding.py:88]: > Python version 3.10 does not support starred expressions in subscriptions
Tuple[*perm]

File: tests/test_binding.py [hunk 3]
```diff
@@ -84,8 +79,10 @@ def factory() -> GetterIntImpl:
 
     # ===========================================
     # clean `_tp_cache`
-    lru_cache_size = 128
-    _ = [GetterInterface[i] for i in range(lru_cache_size)]  # type: ignore[valid-type]
+    from typing import _cleanups as cache_cleanups  # type: ignore[attr-defined]
+
+    for cache_cleanup in cache_cleanups:
+        cache_cleanup()
```
Comment (adriangb) [inline tests/test_binding.py:85]: 👍🏻

File: tests/test_binding.py (hunk 2 above)
Comment (maxzhenzhera) [reply tests/test_binding.py]: The last commit uses [private typing var](https://github.com/python/cpython/blob/edf701c4d83f399bff8612c34fef95da7663945e/Lib/typing.py#L245-L262).
What do you about it?
Looks like it is ok to use it here.

Comment (adriangb) [reply tests/test_binding.py]: I think it's fine

**Comment by maxzhenzhera** [comment]
If we will drop support for python3.7 - is there a need to make a fix using typing_extensions in this PR?

**Comment by adriangb** [comment]
let's keep them independent

**Review by adriangb — APPROVED** [review]

**Pull request merged**